Risco flaky sensor test

<!-- READ THIS FIRST:
  - If you need additional help with this template, please refer to https://www.home-assistant.io/help/reporting_issues/
  - Make sure you are running the latest version of Home Assistant before reporting an issue: https://github.com/home-assistant/core/releases
  - Do not report issues for integrations if you are using custom components or integrations.
  - Provide as many details as possible. Paste logs, configuration samples and code into the backticks.
  DO NOT DELETE ANY TEXT from this template! Otherwise, your issue may be closed without comment.
-->
## The problem
<!-- 
  Describe the issue you are experiencing here to communicate to the
  maintainers. Tell us what you were trying to do and what happened.
-->
- The `test_setup` risco sensor platform test is flaky.
- The problem seems to be that the sensor platform depends on the binary sensor entity state but when the sensor platform is set up and the sensor state is collected, the binary sensor may not be ready yet. The test doesn't take this into account although the production code handles this case with a check.


## Environment
<!--
  Provide details about the versions you are using, which helps us to reproduce
  and find the issue quicker. Version information is found in the
  Home Assistant frontend: Configuration -> Info.
-->

- Home Assistant Core release with the issue: 2021.2.0.dev0
- Last working Home Assistant Core release (if known): 
- Operating environment (OS/Container/Supervised/Core): 
- Integration causing this issue: risco
- Link to integration documentation on our website: 

## Problem-relevant `configuration.yaml`
<!--
  An example configuration that caused the problem for you. Fill this out even
  if it seems unimportant to you. Please be sure to remove personal information
  like passwords, private URLs and other credentials.
-->

```yaml

```

## Traceback/Error logs
<!--
  If you come across any trace or error logs, please provide them.
-->

```txt
______________________________ test_setup[pyloop] ______________________________
[gw0] linux -- Python 3.8.2 /__w/core/core/venv/bin/python

hass = <homeassistant.core.HomeAssistant object at 0x7fd4b42704c0>
two_zone_alarm = <MagicMock id='140551528029344'>

    async def test_setup(hass, two_zone_alarm):  # noqa: F811
        """Test entity setup."""
        registry = await hass.helpers.entity_registry.async_get_registry()
    
        for id in ENTITY_IDS.values():
            assert not registry.async_is_registered(id)
    
        with patch(
            "homeassistant.components.risco.RiscoAPI.get_events",
            return_value=TEST_EVENTS,
        ), patch(
            "homeassistant.components.risco.Store.async_save",
        ) as save_mock:
            entry = await setup_risco(hass)
            await hass.async_block_till_done()
            save_mock.assert_awaited_once_with(
                {LAST_EVENT_TIMESTAMP_KEY: TEST_EVENTS[0].time}
            )
    
        for id in ENTITY_IDS.values():
            assert registry.async_is_registered(id)
    
        for category, entity_id in ENTITY_IDS.items():
>           _check_state(hass, category, entity_id)

tests/components/risco/test_sensor.py:189: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hass = <homeassistant.core.HomeAssistant object at 0x7fd4b42704c0>
category = 'Alarm', entity_id = 'sensor.risco_test_site_name_alarm_events'

    def _check_state(hass, category, entity_id):
        event_index = CATEGORIES_TO_EVENTS[category]
        event = TEST_EVENTS[event_index]
        state = hass.states.get(entity_id)
        assert state.state == event.time
        assert state.attributes["category_id"] == event.category_id
        assert state.attributes["category_name"] == event.category_name
        assert state.attributes["type_id"] == event.type_id
        assert state.attributes["type_name"] == event.type_name
        assert state.attributes["name"] == event.name
        assert state.attributes["text"] == event.text
        assert state.attributes["partition_id"] == event.partition_id
        assert state.attributes["zone_id"] == event.zone_id
        assert state.attributes["user_id"] == event.user_id
        assert state.attributes["group"] == event.group
        assert state.attributes["priority"] == event.priority
        assert state.attributes["raw"] == event.raw
        if event_index == 2:
>           assert state.attributes["zone_entity_id"] == "binary_sensor.zone_1"
E           KeyError: 'zone_entity_id'

tests/components/risco/test_sensor.py:161: KeyError
```

## Additional information

--- a/homeassistant/components/risco/__init__.py
+++ b/homeassistant/components/risco/__init__.py
@@ -60,10 +60,16 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
         EVENTS_COORDINATOR: events_coordinator,
     }
 
-    for component in PLATFORMS:
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component)
+    async def start_platforms():
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_setup(entry, component)
+                for component in PLATFORMS
+            ]
         )
+        await events_coordinator.async_refresh()
+
+    hass.async_create_task(start_platforms())
 
     return True
 

--- a/homeassistant/components/risco/sensor.py
+++ b/homeassistant/components/risco/sensor.py
@@ -73,7 +73,6 @@ class RiscoSensor(CoordinatorEntity):
         self.async_on_remove(
             self.coordinator.async_add_listener(self._refresh_from_coordinator)
         )
-        await self.coordinator.async_request_refresh()
 
     def _refresh_from_coordinator(self):
         events = self.coordinator.data

--- a/tests/components/risco/test_alarm_control_panel.py
+++ b/tests/components/risco/test_alarm_control_panel.py
@@ -167,7 +167,7 @@ async def _check_state(hass, alarm, property, state, entity_id, partition_id):
 
 async def test_states(hass, two_part_alarm):
     """Test the various alarm states."""
-    await setup_risco(hass, CUSTOM_MAPPING_OPTIONS)
+    await setup_risco(hass, [], CUSTOM_MAPPING_OPTIONS)
 
     assert hass.states.get(FIRST_ENTITY_ID).state == STATE_UNKNOWN
     for partition_id, entity_id in {0: FIRST_ENTITY_ID, 1: SECOND_ENTITY_ID}.items():
@@ -249,7 +249,7 @@ async def _call_alarm_service(hass, service, entity_id, **kwargs):
 
 async def test_sets_custom_mapping(hass, two_part_alarm):
     """Test settings the various modes when mapping some states."""
-    await setup_risco(hass, CUSTOM_MAPPING_OPTIONS)
+    await setup_risco(hass, [], CUSTOM_MAPPING_OPTIONS)
 
     registry = await hass.helpers.entity_registry.async_get_registry()
     entity = registry.async_get(FIRST_ENTITY_ID)
@@ -275,7 +275,7 @@ async def test_sets_custom_mapping(hass, two_part_alarm):
 
 async def test_sets_full_custom_mapping(hass, two_part_alarm):
     """Test settings the various modes when mapping all states."""
-    await setup_risco(hass, FULL_CUSTOM_MAPPING)
+    await setup_risco(hass, [], FULL_CUSTOM_MAPPING)
 
     registry = await hass.helpers.entity_registry.async_get_registry()
     entity = registry.async_get(FIRST_ENTITY_ID)
@@ -309,7 +309,7 @@ async def test_sets_full_custom_mapping(hass, two_part_alarm):
 
 async def test_sets_with_correct_code(hass, two_part_alarm):
     """Test settings the various modes when code is required."""
-    await setup_risco(hass, {**CUSTOM_MAPPING_OPTIONS, **CODES_REQUIRED_OPTIONS})
+    await setup_risco(hass, [], {**CUSTOM_MAPPING_OPTIONS, **CODES_REQUIRED_OPTIONS})
 
     code = {"code": 1234}
     await _test_service_call(
@@ -351,7 +351,7 @@ async def test_sets_with_correct_code(hass, two_part_alarm):
 
 async def test_sets_with_incorrect_code(hass, two_part_alarm):
     """Test settings the various modes when code is required and incorrect."""
-    await setup_risco(hass, {**CUSTOM_MAPPING_OPTIONS, **CODES_REQUIRED_OPTIONS})
+    await setup_risco(hass, [], {**CUSTOM_MAPPING_OPTIONS, **CODES_REQUIRED_OPTIONS})
 
     code = {"code": 4321}
     await _test_no_service_call(

--- a/tests/components/risco/test_sensor.py
+++ b/tests/components/risco/test_sensor.py
@@ -1,4 +1,5 @@
 """Tests for the Risco event sensors."""
+from datetime import timedelta
 from unittest.mock import MagicMock, patch
 
 from homeassistant.components.risco import (
@@ -6,12 +7,13 @@ from homeassistant.components.risco import (
     CannotConnectError,
     UnauthorizedError,
 )
-from homeassistant.components.risco.const import DOMAIN, EVENTS_COORDINATOR
+from homeassistant.components.risco.const import DOMAIN
+from homeassistant.util import dt
 
 from .util import TEST_CONFIG, setup_risco
 from .util import two_zone_alarm  # noqa: F401
 
-from tests.common import MockConfigEntry
+from tests.common import MockConfigEntry, async_fire_time_changed
 
 ENTITY_IDS = {
     "Alarm": "sensor.risco_test_site_name_alarm_events",
@@ -176,7 +178,7 @@ async def test_setup(hass, two_zone_alarm):  # noqa: F811
     ), patch(
         "homeassistant.components.risco.Store.async_save",
     ) as save_mock:
-        entry = await setup_risco(hass)
+        await setup_risco(hass)
         await hass.async_block_till_done()
         save_mock.assert_awaited_once_with(
             {LAST_EVENT_TIMESTAMP_KEY: TEST_EVENTS[0].time}
@@ -188,14 +190,13 @@ async def test_setup(hass, two_zone_alarm):  # noqa: F811
     for category, entity_id in ENTITY_IDS.items():
         _check_state(hass, category, entity_id)
 
-    coordinator = hass.data[DOMAIN][entry.entry_id][EVENTS_COORDINATOR]
     with patch(
         "homeassistant.components.risco.RiscoAPI.get_events", return_value=[]
     ) as events_mock, patch(
         "homeassistant.components.risco.Store.async_load",
         return_value={LAST_EVENT_TIMESTAMP_KEY: TEST_EVENTS[0].time},
     ):
-        await coordinator.async_refresh()
+        async_fire_time_changed(hass, dt.utcnow() + timedelta(seconds=65))
         await hass.async_block_till_done()
         events_mock.assert_awaited_once_with(TEST_EVENTS[0].time, 10)
 

--- a/tests/components/risco/test_sensor.py
+++ b/tests/components/risco/test_sensor.py
@@ -1,6 +1,6 @@
 """Tests for the Risco event sensors."""
 from datetime import timedelta
-from unittest.mock import MagicMock, patch
+from unittest.mock import MagicMock, PropertyMock, patch
 
 from homeassistant.components.risco import (
     LAST_EVENT_TIMESTAMP_KEY,
@@ -10,7 +10,7 @@ from homeassistant.components.risco import (
 from homeassistant.components.risco.const import DOMAIN
 from homeassistant.util import dt
 
-from .util import TEST_CONFIG, setup_risco
+from .util import TEST_CONFIG, TEST_SITE_UUID, setup_risco
 from .util import two_zone_alarm  # noqa: F401
 
 from tests.common import MockConfigEntry, async_fire_time_changed
@@ -172,23 +172,28 @@ async def test_setup(hass, two_zone_alarm):  # noqa: F811
     for id in ENTITY_IDS.values():
         assert not registry.async_is_registered(id)
 
+    await setup_risco(hass)
+    await hass.async_block_till_done()
+    for id in ENTITY_IDS.values():
+        assert registry.async_is_registered(id)
+
     with patch(
         "homeassistant.components.risco.RiscoAPI.get_events",
         return_value=TEST_EVENTS,
     ), patch(
+        "homeassistant.components.risco.RiscoAPI.site_uuid",
+        new_callable=PropertyMock(return_value=TEST_SITE_UUID),
+    ), patch(
         "homeassistant.components.risco.Store.async_save",
     ) as save_mock:
-        await setup_risco(hass)
+        async_fire_time_changed(hass, dt.utcnow() + timedelta(seconds=65))
         await hass.async_block_till_done()
+
         save_mock.assert_awaited_once_with(
             {LAST_EVENT_TIMESTAMP_KEY: TEST_EVENTS[0].time}
         )
-
-    for id in ENTITY_IDS.values():
-        assert registry.async_is_registered(id)
-
-    for category, entity_id in ENTITY_IDS.items():
-        _check_state(hass, category, entity_id)
+        for category, entity_id in ENTITY_IDS.items():
+            _check_state(hass, category, entity_id)
 
     with patch(
         "homeassistant.components.risco.RiscoAPI.get_events", return_value=[]

--- a/tests/components/risco/test_sensor.py
+++ b/tests/components/risco/test_sensor.py
@@ -172,22 +172,15 @@ async def test_setup(hass, two_zone_alarm):  # noqa: F811
     for id in ENTITY_IDS.values():
         assert not registry.async_is_registered(id)
 
-    await setup_risco(hass)
-    await hass.async_block_till_done()
-    for id in ENTITY_IDS.values():
-        assert registry.async_is_registered(id)
-
     with patch(
-        "homeassistant.components.risco.RiscoAPI.get_events",
-        return_value=TEST_EVENTS,
-    ), patch(
         "homeassistant.components.risco.RiscoAPI.site_uuid",
         new_callable=PropertyMock(return_value=TEST_SITE_UUID),
     ), patch(
         "homeassistant.components.risco.Store.async_save",
     ) as save_mock:
-        async_fire_time_changed(hass, dt.utcnow() + timedelta(seconds=65))
-        await hass.async_block_till_done()
+        await setup_risco(hass, TEST_EVENTS)
+        for id in ENTITY_IDS.values():
+            assert registry.async_is_registered(id)
 
         save_mock.assert_awaited_once_with(
             {LAST_EVENT_TIMESTAMP_KEY: TEST_EVENTS[0].time}

--- a/tests/components/risco/util.py
+++ b/tests/components/risco/util.py
@@ -17,7 +17,7 @@ TEST_SITE_UUID = "test-site-uuid"
 TEST_SITE_NAME = "test-site-name"
 
 
-async def setup_risco(hass, options={}):
+async def setup_risco(hass, events=[], options={}):
     """Set up a Risco integration for testing."""
     config_entry = MockConfigEntry(domain=DOMAIN, data=TEST_CONFIG, options=options)
     config_entry.add_to_hass(hass)
@@ -33,6 +33,9 @@ async def setup_risco(hass, options={}):
         new_callable=PropertyMock(return_value=TEST_SITE_NAME),
     ), patch(
         "homeassistant.components.risco.RiscoAPI.close"
+    ), patch(
+        "homeassistant.components.risco.RiscoAPI.get_events",
+        return_value=events,
     ):
         await hass.config_entries.async_setup(config_entry.entry_id)
         await hass.async_block_till_done()

